orchagent/orchdaemon.cpp

```diff
@@ -313,5 +358,81 @@ void OrchDaemon::start()
         for (Orch *o : m_orchList)
             o->doTask();
 
+        /*
+         * All data to be restored have been added to m_toSync of each orch
+         * at contructor phase.  And the order of m_orchList guranteed the
+         * dependency of tasks had been met, restore is done.
+         */
+        if (!restored && m_select->isQueueEmpty() && gPortsOrch->isInitDone())
+        {
+            /*
+             * drain remaining data that are out of order like LAG_MEMBER_TABLE and VLAN_MEMBER_TABLE
+             * since they were checked before LAG_TABLE and VLAN_TABLE.
+             */
+            for (Orch *o : m_orchList)
+            {
+                o->doTask();
+            }
+
+            warmRestoreValidation();
+            SWSS_LOG_NOTICE("Orchagent state restore done");
+            restored = true;
+            syncd_apply_view();
+
+            // TODO: should be set after port/fdb/arp sync up
+            WarmStart::setWarmStartState("orchagent", WarmStart::RECONCILED);
+
+            /* Pick up those tasks postponed by restore processing */
+            if(!executorSet.empty())
+            {
+                for (Executor *c : executorSet)
+                {
+                    c->execute();
+                }
+                for (Orch *o : m_orchList)
+                {
+                    o->doTask();
+                }
+            }
+        }
+    }
+}
+
+/*
+ * Get tasks to sync for consumers of each orch being managed by this orch daemon
+ */
+void OrchDaemon::getTaskToSync(vector<string> &ts)
+{
+    for (Orch *o : m_orchList)
+    {
+        o->dumpTasks(ts);
+    }
+}
+
+
+/* Perform basic validation after start restore for warm start */
+bool OrchDaemon::warmRestoreValidation()
+{
+    /*
+     * No pending task should exist for any of the consumer at this point.
+     * All the prexisting data in appDB and configDb have been read and processed.
+     */
+    vector<string> ts;
+    getTaskToSync(ts);
+    if (ts.size() != 0)
+    {
+        // TODO: change it to fatal once staged ProducerStateTable/ConsumerStateTable change and
```

Comment:
remove TODO here?

Reply:
For planned warm restart,  there is one PR https://github.com/Azure/sonic-swss/pull/562/files     for enforcing orchagent to be in certain state before stopping orchagent/swss docker.  After orchagent is restored, it should gain back the same state.

But there is open issue about the exact criteria of warm restart readiness. Currently we want to make sure that there is no un-met object dependency in orchagent which means the "SyncMap m_toSync;" of each Consumer is empty.   @zhenggen-xu  ever expressed concern over this check.

Reply:
This check is not necessary reach the desired result.  The problem is if some dependencies were never met (e,g, the nexthop is not resolved for some reason for a particular route), which was the case for pre-reboot,  but may kill the whole warm reboot process. Looks like we may want to handle those entries differently or we simply do a loose check (print warning etc) and move forward.

Reply:
Do we want to allow this scenario (there is unfulfilled dependency) for planned warm restart?

The possible alternative is to add a knob to have two levels of pre-reboot check:
1. Allow unfulfilled dependency.  
2. disallow unfulfilled dependency.  
After warm restore, we do the check accordingly.

With no intention to make it a hot topic here and block everything else,  I'll change the log level to NOTICE and continue for now.

Please go to https://github.com/Azure/sonic-swss/pull/562 for further discussion.